### PR TITLE
Building Improvement

### PR DESCRIPTION
**Pull request by Aceship**
**Reference Plane for building**
### 
You can start using reference plane with 
- left grip and pointing to some building parts

Reference plane allows you to easily snap to where you can, this helps much when making a floating floor and expanding walls 

Reference plane orientation can be scrolled by using right analog stick up and down with 3 types : 
- Vertical (force the reference plane aligned vertically) (helps when making a bridge support downwards)
   (indicated with vertical ring)
- Normal (aligned to the building part that pointed on) 
   (indicated with no ring)
- Horizontal (force the reference plane aligned horizontally) (helps when making floor after a stair) 
   (indicated with horizontal ring)

**Other Building Stuff**
###
- Add option to do build on releasing button (enabled by default for reference plane)
- Allow rotate towards/away from hand direction by using Right hand grip + analog up/down

**Snapping Tool**
You can use dedicated snapping tool by hold right trigger while pointing on building parts (it will stick to the same location until done)

especially useful when making bridge support

**Advance building mode (Off by default, can be enabled in option)**
Free/Precision mode
By holding grab on the ring near the hammer, the placement mode changed, and the placement ghost can be moved with :

- left+right grab to move the building piece, it follows both position & rotation of both hands

- right hand grabbing the gizmo that follow the left hand will move the piece following the axis its grabbed to

Advance rotation gizmo 
- Left hand grabbing the gizmo that follow the right hand (the one with ring) will rotate the piece, works like crank , with angle snap, rotating further from the center allows more precision (also works while not in free/precision mode)

There are 2 types of advanced rotation gizmo , normal and world axis 
can be changed by grabbing the lower end of the hammer till it change color

while in normal mode (yellow) rotation axis will follow the model 
while in world axis mode (cyan) , the rotation vertical axis will follow the world vertical axis  while the x and z axis will try to follow the object forward 

Advanced Rotation with analog

- grab with right analog stick, left and right will rotate the piece , rotate around vertically (can be changed in config to rotate based on the building part rotation)

- grab with right analog stick, up and down will reset all 3 axis and follow the rotation towards the controller direction, if holding it long enough, it will follow the rotation of the parts currently aimed

### CODE REVIEW

File: ValheimVRMod/Scripts/BuildingManager.cs [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+            var onHandPoints = onHand.transform;
+            if (!onHandPoints)
+            {
+                return;
+            }
+            if (!pieceRaycast.transform)
+            {
+                return;
+            }
+            var aimedPoints = pieceRaycast.parent.transform ? pieceRaycast.parent.transform : pieceRaycast;
```
Comment (brandonmousseau) [inline ValheimVRMod/Scripts/BuildingManager.cs:347]: pieceRaycast.parent returns the parent transform of pieceRaycast, so the ".transform" part is redundant here I believe and may actually through a null reference is parent is null

Comment (Aceship) [reply ValheimVRMod/Scripts/BuildingManager.cs]: yeah apparently thats the problem, thanks 👍

File: ValheimVRMod/VRCore/UI/VRHud.cs [hunk 2]
```diff
@@ -315,9 +349,22 @@ private bool ensureHudCanvas()
             rightHudCanvas.transform.localRotation = Quaternion.identity;
             rightHudCanvas.worldCamera = hudCamera;
 
+            buildHudCanvas = buildHudCanvasParent.AddComponent<Canvas>();
```
Comment (Andi1986) [inline ValheimVRMod/VRCore/UI/VRHud.cs:352]: we should extract this into a method that accepts a canvas, canvasgroup and canvasParaent as parameter

File: ValheimVRMod/Scripts/BuildingManager.cs [hunk 3]
```diff
@@ -0,0 +1,1286 @@
...
+        private RaycastHit lastRefCast;
+        private GameObject buildRefBox;
+        private GameObject buildRefPointer;
+        private GameObject buildRefPointer2;
+        private int currRefType;
+        private bool refWasChanged;
+        public static BuildingManager instance;
+
+        //Snapping stuff
+        private static bool isSnapping = false;
```
Comment (Andi1986) [inline ValheimVRMod/Scripts/BuildingManager.cs:25]: remove these static modifiers please.
Ideally, only the instance variable should be static.
i had this bad static behaviour aswell at the beginning

Comment (Aceship) [reply ValheimVRMod/Scripts/BuildingManager.cs:25]: so its supposedly use `BuildingManager.instance.(function)` instead ?

File: ValheimVRMod/Patches/UIPatches.cs [hunk 4]
```diff
@@ -810,5 +810,24 @@ private static void clearHotbar(HotkeyBar hotkeyBar)
             hotkeyBar.m_elements.Clear();
         }
     }
+
+    [HarmonyPatch(typeof(Hud),nameof(Hud.UpdateStamina))]
+    class StaminaPatch
+    {
+        private static void Postfix(RectTransform ___m_staminaBar2Root, float ___m_staminaHideTimer, Animator ___m_staminaAnimator)
+        {
+            if (!VHVRConfig.NonVrPlayer() && VHVRConfig.StaminaPanelPlacement() != "Legacy" && !VHVRConfig.UseLegacyHud())
```
Comment (Andi1986) [inline ValheimVRMod/Patches/UIPatches.cs:819]: use the nonVrPlayer check as early return (also since second if statement below ignores it)

**Review by brandonmousseau — APPROVED** [review]

File: ValheimVRMod/VRCore/UI/VRControls.cs [hunk 5]
```diff
@@ -152,16 +153,27 @@ private void updateAltMapZoomTimer()
             bool rightClickUp = false;
             if (useRightClick && laserControlsActive && inPlaceMode())
             {
-                rightClickDown = SteamVR_Actions.laserPointers_RightClick.GetStateDown(SteamVR_Input_Sources.Any);
+                rightClickDown = SteamVR_Actions.laserPointers_RightClick.GetState(SteamVR_Input_Sources.Any);
                 rightClickUp = SteamVR_Actions.laserPointers_RightClick.GetStateUp(SteamVR_Input_Sources.Any);
+                if(rightClickDown)
+                    buildQuickActionTimer += Time.unscaledDeltaTime;
```
[the remark this answers is not in the document]
Comment (Aceship) [reply ValheimVRMod/VRCore/UI/VRControls.cs:159]: its for delaying quick action from appearing when in build mode, making sure it doesnt accidentally change equip when you wanna change building piece

iirc im following some of the other like zoom and rotation above it, that use `unscaledDeltaTime`, was thinking it should use that too

File: ValheimVRMod/VRCore/UI/VRControls.cs [hunk 6]
```diff
@@ -409,6 +421,55 @@ public float GetJoyRightStickY()
...
+                return 1;
+            }
+            else
+            {
+                return 0;
+            }
+        }
+        public int getDirectRightXAxis()
+        {
+            float yAxis = -pitchAndYaw.axis.x;
```
[the remark this answers is not in the document]
Comment (Aceship) [reply ValheimVRMod/VRCore/UI/VRControls.cs]: changed 👍

File: ValheimVRMod/Patches/UIPatches.cs [hunk 7]
```diff
@@ -810,5 +810,27 @@ private static void clearHotbar(HotkeyBar hotkeyBar)
             hotkeyBar.m_elements.Clear();
         }
     }
+
+    [HarmonyPatch(typeof(Hud),nameof(Hud.UpdateStamina))]
+    class StaminaPatch
+    {
+        private static void Postfix(RectTransform ___m_staminaBar2Root, float ___m_staminaHideTimer, Animator ___m_staminaAnimator)
+        {
+            if (VHVRConfig.NonVrPlayer())
```
[the remark this answers is not in the document]
Comment (Aceship) [reply ValheimVRMod/Patches/UIPatches.cs]: yeah i guess its not used in KB&M

changed 👍

**Pull request merged**